pkg/admin/prerun/prerun.go

```diff
@@ -109,6 +115,109 @@ func (pr *PreRun) Perform() error {
+// determines if regular pre-run procedure should be perfomed.
+//
+// It will also handle special scenarios such as: upgrade from 4.13.
+func (pr *PreRun) checkPreconditions() (bool, error) {
+	dataExists, err := util.PathExists(config.DataDir)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if data directory already exists: %w", err)
+	}
+
+	versionExists, err := util.PathExists(versionFilePath)
+	if err != nil {
+		return false, fmt.Errorf("checking if version metadata exists failed: %w", err)
+	}
+
+	healthExists, err := util.PathExists(healthFilepath)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if health file already exists: %w", err)
+	}
+
+	klog.InfoS("Existence of important paths",
+		"data-exists?", dataExists,
+		"version-exists?", versionExists,
+		"health-exists?", healthExists,
+	)
+
+	/*
+		| id  | data | version | health | comment                                                                                                     |
+		| --- | ---- | ------- | ------ | ----------------------------------------------------------------------------------------------------------- |
+		| 1   | 0    | 0       | 0      | first, clean start of MicroShift                                                                            |
+		| 2   | 0    | 0       | 1      | data removed manually, but health/backups kept                                                              |
+		| 3   | 0    | 1       | 0      | impossible to detect right now [0]                                                                          |
+		| 4   | 0    | 1       | 1      | impossible to detect right now [0]                                                                          |
+		| 5   | 1    | 0       | 0      | upgrade from 4.13                                                                                           |
+		| 6   | 1    | 0       | 1      | first boot failed very early, or upgrade from 4.13 failed (e.g. healthcheck didn't see service being ready) |
+		| 7   | 1    | 1       | 0      | first start, rebooted before green/red scripts managed to write health info                                 |
+		| 8   | 1    | 1       | 1      | regular                                                                                                     |
+
+		[0] it would need a comprehensive check if data exists, not just existence of /var/lib/microshift
+	*/
+
+	if !dataExists {
+		// 1
+		if !healthExists {
+			klog.InfoS("Neither data dir nor health file exist - assuming first start of MicroShift")
+			return false, nil
+		}
+
+		// 2
+		// TODO: Health needs to be extended into a history of boots and their health
+		// so prerun can look into the past and decide if a backup should be restored
+		// for currently running deployment
+		return false, fmt.Errorf("TODO: data is missing, but health file exists")
+	}
+
+	// 5
+	if !versionExists && !healthExists {
```

Comment:
It would be good to be consistent about either having each case be 1 `if` statement with combined conditions (like here) or to use nested `if` statements (like above). I don't have a strong preference for which approach is used, but if nested conditions make it easier to ensure we are testing all of the relevant cases maybe that's the thing to do?

Reply:
What I like about current version is clear distinction of cases where data does and does not exist (seems to me to be kind of a big deal that deserves special treatment).
Lines from 174 to the end of the function could be inside an `else` but that's not idiomatic...
TBH I'm most happy with the current version - it's explicit that `if !dataExists` exhausts all the cases when data doesn't exist, and the rest are cases when it does. Turning that into `if !dataExists && !healthExists` and `if !dataExists && healthExists` makes me spend some brain cycles to think if there are other, implicit cases with `!dataExists`